engine/table/src/main/java/io/deephaven/engine/table/impl/hierarchical/RollupTableImpl.java

```diff
@@ -451,6 +473,51 @@ protected RollupTableImpl copy() {
                 levelTables, levelRowLookups, levelNodeTableSources,
                 aggregatedNodeDefinition, aggregatedNodeOperations,
                 constituentNodeDefinition, constituentNodeOperations,
+                lowestRollupKeyFilters,
+                availableColumnDefinitions);
+    }
+
+    @Override
+    public RollupTableImpl rebase(final Table newSource) {
+        if (!newSource.getDefinition().equals(source.getDefinition())) {
+            if (newSource.getDefinition().equalsIgnoreOrder(source.getDefinition())) {
+                throw new IllegalArgumentException(
+                        "Cannot rebase a RollupTable with a new source definition, column order is not identical.");
+            }
+            final String differenceDescription = newSource.getDefinition()
+                    .getDifferenceDescription(source.getDefinition(), "new source", "existing source", ",");
+            throw new IllegalArgumentException(
+                    "Cannot rebase a RollupTable with a new source definition: " + differenceDescription);
+        }
+
+        final QueryTable newSourceQueryTable;
+        if (lowestRollupKeyFilters != null) {
+            // if this rollup has had any key filters, then we apply them to the new source rather than generating
+            // buckets for keys which cannot be present in the result. Because these filters are applied, we pass
+            // in a null set of filters to the new RollupTableImpl we are creating.
+            final List<WhereFilter> keyFilter =
+                    Arrays.stream(lowestRollupKeyFilters).map(WhereFilter::copy).collect(Collectors.toList());
+            newSourceQueryTable = (QueryTable) newSource.where(Filter.and(keyFilter));
+        } else {
+            newSourceQueryTable = (QueryTable) newSource;
+        }
+
+        final int numLevels = groupByColumns.size() + 1;
```

Comment:
I would refactor `makeRollup` to have a more complete version that just accepts the node-level parts in order to avoid this duplication. So we would filter the source, then delegate to `makeRollupInternal` or whatever.